src/graph/graph_op.cc

```diff
@@ -315,4 +315,74 @@ ImmutableGraph GraphOp::ToSimpleGraph(const GraphInterface* graph) {
   return ImmutableGraph(csr);
 }
 
+Graph GraphOp::ToBidirectedMutableGraph(const GraphInterface* g) {
+  std::unordered_map<int, std::unordered_map<int, int>> n_e;
+  for (dgl_id_t u = 0; u < g->NumVertices(); ++u) {
+    for (const dgl_id_t v : g->SuccVec(u)) {
+      n_e[u][v]++;
+    }
+  }
+
+  Graph bg;
+  bg.AddVertices(g->NumVertices());
+  for (dgl_id_t u = 0; u < g->NumVertices(); ++u) {
+    for (dgl_id_t v = u; v < g->NumVertices(); ++v) {
+      const auto new_n_e = std::max(n_e[u][v], n_e[v][u]);
+      if (new_n_e > 0) {
+        IdArray us = IdArray::Empty({new_n_e}, DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+        int64_t* us_data = static_cast<int64_t*>(us->data);
+        std::fill(us_data, us_data + new_n_e, u);
+        if (u == v) {
+          bg.AddEdges(us, us);
+        } else {
+          IdArray vs = IdArray::Empty({new_n_e}, DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+          int64_t* vs_data = static_cast<int64_t*>(vs->data);
+          std::fill(vs_data, vs_data + new_n_e, v);
+          bg.AddEdges(us, vs);
+          bg.AddEdges(vs, us);
+        }
+      }
+    }
+  }
+  return bg;
+}
+
+ImmutableGraph GraphOp::ToBidirectedImmutableGraph(const GraphInterface* g) {
+  std::unordered_map<int, std::unordered_map<int, int>> n_e;
+  for (dgl_id_t u = 0; u < g->NumVertices(); ++u) {
+    for (const dgl_id_t v : g->SuccVec(u)) {
+      n_e[u][v]++;
+    }
+  }
+
+  std::vector<dgl_id_t> indptr(g->NumVertices() + 1), indices;
+  indptr[0] = 0;
+  for (dgl_id_t u = 0; u < g->NumVertices(); ++u) {
+    std::unordered_set<dgl_id_t> hashmap;
+    std::vector<dgl_id_t> nbrs;
+    for (const dgl_id_t v : g->SuccVec(u)) {
+      if (!hashmap.count(v)) {
+        nbrs.push_back(v);
+        hashmap.insert(v);
+      }
+    }
+    for (const dgl_id_t v : g->PredVec(u)) {
+      if (!hashmap.count(v)) {
+        nbrs.push_back(v);
+        hashmap.insert(v);
+      }
+    }
+    for (const dgl_id_t v : nbrs) {
+      const auto new_n_e = std::max(n_e[u][v], n_e[v][u]);
+      for (size_t i = 0; i < new_n_e; ++i) {
+        indices.push_back(v);
+      }
+    }
+    indptr[u+1] = indices.size();
+  }
+  CSRPtr csr(new CSR(g->NumVertices(), indices.size(),
+        indptr.begin(), indices.begin(), RangeIter(0), g->IsMultigraph()));
+  return ImmutableGraph(csr);
```

Comment:
The immutable graph supports construction from a COO structure. This should be easier and more efficient.

Reply:
Done.